relays/messages-relay/src/message_race_delivery.rs

```diff
@@ -164,280 +232,87 @@ type MessageDeliveryStrategy<P> = DeliveryStrategy<
 		&mut self,
-		race_state: &RaceState<
-			HeaderId<SourceHeaderHash, SourceHeaderNumber>,
-			HeaderId<TargetHeaderHash, TargetHeaderNumber>,
-			Nonce,
-			Proof,
-		>,
-	) -> Option<RangeInclusive<Nonce>> {
-		// if we have already selected nonces that we want to submit, do nothing
-		if race_state.nonces_to_submit.is_some() {
-			return None;
-		}
-
-		// if we already submitted some nonces, do nothing
-		if race_state.nonces_submitted.is_some() {
-			return None;
-		}
-
-		// 1) we want to deliver all nonces, starting from `target_nonce + 1`
-		// 2) we want to deliver at most `self.max_nonces_to_relay_in_single_tx` nonces in this batch
-		// 3) we can't deliver new nonce until header, that has emitted this nonce, is finalized
-		// by target client
-		let nonces_begin = self.target_nonce + 1.into();
-		let best_header_at_target = &race_state.target_state.as_ref()?.best_peer;
-		let mut nonces_end = None;
-		let mut i = Zero::zero();
-		while i < self.max_nonces_to_relay_in_single_tx {
-			let nonce = nonces_begin + i;
-
-			// if queue is empty, we don't need to prove anything
-			let (first_queued_at, first_queued_nonce) = match self.source_queue.front() {
-				Some((first_queued_at, first_queued_nonce)) => ((*first_queued_at).clone(), *first_queued_nonce),
-				None => break,
-			};
-
-			// if header that has queued the message is not yet finalized at bridged chain,
-			// we can't prove anything
-			if first_queued_at.0 > best_header_at_target.0 {
-				break;
+		race_state: &RaceState<SourceHeaderIdOf<P>, TargetHeaderIdOf<P>, P::MessageNonce, P::MessagesProof>,
+	) -> Option<(RangeInclusive<P::MessageNonce>, bool)> {
+		const CONFIRMED_NONCE_PROOF: &str = "\
+			ClientNonces are crafted by MessageDeliveryRace(Source|Target);\
+			MessageDeliveryRace(Source|Target) always fills confirmed_nonce field;\
+			qed";
+
+		let source_nonces = self.source_nonces.as_ref()?;
+		let target_nonces = self.target_nonces.as_ref()?;
+
+		// There's additional condition in the message delivery race: target would reject messages
+		// if there are too much unconfirmed messages at the inbound lane.
+
+		// https://github.com/paritytech/parity-bridges-common/issues/432
+		// TODO: message lane loop works with finalized blocks only, but we're submitting transactions that
+		// are updating best block (which may not be finalized yet). So all decisions that are made below
+		// may be outdated. This needs to be changed - all logic here bust be built on top of best blocks.
+
+		// The receiving race is responsible to deliver confirmations back to the source chain. So if
+		// there's a lot of unconfirmed messages, let's wait until it'll be able to do its job.
+		let latest_received_nonce_at_target = target_nonces.latest_nonce;
+		let latest_confirmed_nonce_at_source = source_nonces.confirmed_nonce.expect(CONFIRMED_NONCE_PROOF);
+		let confirmatins_missing = latest_received_nonce_at_target.checked_sub(&latest_confirmed_nonce_at_source);
```

Comment:
Is there a problem if we have a problem if `latest_confirmed_nonce_at_source` > `latest_received_nonce_at_target`? This seems like something that shouldn't happen

Reply:
It should not happen, right. In theory :)